README
Summary
=======

This software allows you to download and compile high-resolution topographic maps from the NSW geospatial data servers, covering all of NSW and the ACT. The resulting maps include most of the features found in the printed NSW topographic map series. You can specify the exact extent of the area which you wish to map, as well as your desired print resolution (in pixels per inch) and scale (typically 1:25000 or 1:50000). You can obtain the map output as a single composite file, or a multi-layer file containing layers for each topographic feature (e.g. contours, watercourses, roads, etc). The output map is also georeferenced for use with map-viewing or GIS software.

This software was originally designed for the production of rogaining maps and as such includes several extra features (such as aerial imagery overlays, marker layers for control checkpoints, and magnetic declination lines). However the software is useful for anyone wanting to create custom NSW topo maps for outdoor recreation.

Pre-Requisites
==============

The software is run as a script, so you will need some familiarity with the command line. It was developed on a Mac, should work fine under linux/BSD, and is worth trying with Windows.

You will need to install several open-source tools before running the script:
- The Ruby programming language (http://ruby-lang.org TODO). You'll need the more recent Ruby 1.9.x, not 1.8.x. On a Mac OS, Ruby should come with XCode when you install it, or you could use MacPorts. On Linux, just install Ruby using your system's package manager. On Windows, a complete Ruby installation can be downloaded from (TODO).
- ImageMagick, a command-line image manipulation tool (http://imagemagick.org). Again, install using MacPorts on Mac OS, your Linux package manager, or download a pre-built binary (TODO) for Windows.
- The GDAL command-line utilities (http://gdal.org TODO). These are utilities for processing geospatial raster data. The tools are available as a standalone package. You can also get them as part of QGIS (TODO), which you may wish to install if you need GIS software for viewing the output maps and GPS waypoints and tracks.
- If you plan to make further enhancements, manual corrections or additions to the map, a layer-based image editing tool such as Gimp (open-source) or Photoshop (proprietary) is also needed.

You can check that the tools are correctly installed by using the following commands:
- ruby -v
- identify -version
- gdalwarp --version
You should receive version information for each tool if it is installed correctly and in your path.

2Gb should be considered the minimum memory requirement, and more is always better. (With only 2Gb, larger maps will start paging your memory to disk during the compositing steps.) You will also need a decent internet connection. Most of the topographic map layers won't use a lot of bandwidth, but the aerial imagery could amount to 100Mb or more for a decent-sized map. (And don't bother with dialup!)

Usage
=====

You will first need to create a directory for the map you are building. Running the script will result in a large number of image files representing the map layers, so a directory is needed to contain them.

In this directory, create and edit a configuration text file called 'config.yml' which contains the bounds of the area you want mapped. (This format of this file is YAML [TODO wikipedia link], though you don't really need to know this.) Specify the map bounds by providing minimum and maximum coordinates in each direction, as follows:

zone: 55
eastings:
  - 730500
  - 741500
northings:
  - 6014500
  - 6022500

or,

latitudes: 
  - -35.951221
  - -35.892871
longitudes: 
  - 149.383789
  - 149.489746

(When specifying coordinates in UTM you must also specify the UTM zone, which in NSW will be 54, 55 or 56.)

Alternatively, you can specify a single coordinate for the map's centre, and a physical size for the map. The map size should be in units of mm, cm or in. For example:

zone: 55
easting: 691750
northing: 6070500
size: 22cm x 36cm

or,

latitude: -33.474050497749076
longitude: 150.13797998428345
size: 6cm x 6cm

Once you have created your configuration file, run the script to create your map. The script itself is the 'nswtopo.rb' file. The easiest way is to copy this file into your folder and run it from there thusly: 'ruby nswtopo.rb'. Alternatively, keep the script elsewhere and run it as 'ruby /path/to/nswtopo.rb'. By giving the script exec priveleges ('chmod +x nswtopo.rb' or equivalent), you can run it directly ('./nswtopo.rb'; you may need to modify the hash-bang on line 1 to reflect the location of your ruby binary).

When the script starts it will list the scale of your map (e.g. 1:25000), its physical size and resolution (e.g. 38cm x 24cm @ 300 ppi) and its size in megapixels. For a 300 pixel-per-inch image (the default), an A3 map should be about 15 megapixels. An unexpectedly large or small number may indicate an error in your configuration file; similarly, if all the topographic layers are missed, this probably indicates you've incorrectly specified bounds outside NSW. You can ctrl-c at any point to stop the script; it will pick up where it left off the next time you run it.

The script will then proceed to download a large number of layers. A progress bar will show for each layer. Depending on your connection and the size of your map, an hour or more may be required. (I suggest starting with a small map, say 8cm x 8cm, just to familiarize yourself with the software; this should only take a few minutes.) Any errors received will be displayed and the layer skipped; you can run the script again to retry the skipped layers as they are usually just temporary server errors. (Layers already present will not be retried.)

A description of each layer is at the end of this document.

After all layers have been downloaded, the script will then compile them into a final map image. The default is to create both a PNG and a multi-layered TIFF. (Several output formats are possible - see below). Depending on the specs of your computer, creating a multi-layered TIFF or PSD (photoshop) image may take a long time, particularly if swap memory is hit. Be patient. If you don't plan to do any further manipulation or editing of your map, just specify a PNG or PDF as output format, as this will take less time.

Using the Output
================

You can use the output files in a few different ways. If you just want a quick map, specify your output format as png or pdf only. The program will automatically compile all the data into a single file which you can then print and use.

If you're creating a map for rogaining, you will probably want to build a multi-layered file for further editing using Gimp or Photoshop. In this case, specify 'layered.tif' or 'psd', respectively, as your output format. These formats will keep each topographic feature on a separate layer, allowing you to edit them individually. In combination with the aerial layers, which you can turn on and off as underlayers, this allows you to compare the mapped location of topographic features such as roads, dams and cliffs against their position on the aerial imagery, and to manually add, correct or remove such features as needed (e.g. old firetrails that have been changed, are no longer present, or new firetrails that have not yet been mapped).

(Note that the ImageMagick photoshop driver is not particularly good - it does not do any file compression, which can lead to a gigabyte+ file size and very slow performance during creation of the psd file. The file will be correctly compressed once you load and save it in Photoshop, however.)

It is also possible to construct your own Photoshop or Gimp document by hand, using the topographic layers as layer masks for color fill layers representing each feature. The topographic feature layers are colored white-on-black to allow you to do this easily.

Map Configuration
=================

By editing the 'config.yml' file you can customise a number of aspects of your map, including the colour and patterns used for various features and which layers to include. If no other configuration is provided, reasonable defaults are used, as shown below:

# Set the scale and print resolution of the map. 300-400 ppi is probably optimal for most maps, and
# gives a resolution of about 2.1 metres per pixel at 1:25000. Going beyond 400 ppi will not yield any
# more detail, will make the downloads slower and will blow out the megapixel count considerably.

scale: 25000              # desired map scale (1:25000 in this case)
ppi: 300                  # print resolution in pixels per inch

# Set the filename for the output map and related files.

name: map                 # filename to use for the final map

# Specify the contour spacing. The standard contour coverage is 10m for eastern NSW and 20m for central
# and western NSW (and most of the Snowy Mountains, disappointingly). Large scale contour data
# (source: 2) at 1m or 2m intervals seems limited to coastal areas (e.g. Merimbula), and is probably not
# of much use as it does not match up with watercourse features very well.

contours:
  interval: 10            # contour interval in metres
  index: 100              # index contour interval in metres
  labels: 50              # interval in metres for contour labels
  source: 1               # elevation capture program: 1 for medium scale (1:25000 - 1:100000);
                          # 2 for large scale (1:1000 - 1:10000); 4 for contours derived from DEM.

# A layer containing lines of magnetic declination is automatically produced. The magnetic declination
# angle for the map centre is automatically retrieved from the Geoscience Australia website.

declination:
  spacing: 1000           # perpendicular spacing of magnetic declination lines in metres

# If the map is for general-purpose use, a UTM grid is generated for use with a GPS.

grid:
  intervals:              # horizontal and vertical spacing of UTM grid lines in metres
    - 1000                # (East-West grid spacing)
    - 1000                # (North-South grid spacing)
  fontsize: 6.0           # font size of UTM grid labels
  family: Arial Narrow    # font family of UTM grid labels
  weight: 200             # font weight of UTM grid labels

# Shaded relief and elevation layers are automatically produced from the ASTER digital elevation
# model.
  
relief:
  azimuth:                # azimuth angle for shaded relief layers (degrees clockwise from North)
    - 315                 # (315 degrees is the standard shaded relief angle)
    - 45                  # (each azimuthal angle produces a separate shaded relief layer)
  altitude: 45            # angle of illumination from horizon (45 is standard)
  exaggeration: 1         # vertical exaggeration factor

# Drop in you control waypoint file (in .gpx or .kml format) and layers containing control circles
# and numbers will be automatically generated. If a waypoint is name 'HH' it will be drawn as a
# triangle, otherwise a circle will be drawn.

controls:
  file: controls.gpx      # filename (.gpx or .kml format) of control waypoint file
  fontsize: 14            # font size for control numbers
  diameter: 7.0           # diameter of control circles in millimetres
  thickness: 0.2          # thickness of control circles in millimetres

# Specify the format(s) of the output map files you would like to create. Choose from 'png',
# 'tif', 'gif', 'bmp', 'pdf', 'psd' and 'layered.tif'. TIFF files will be automatically
# georeferenced with geotiff tags for use with GIS software. If you specify PNG/GIF/BMP format,
# a '.map' file will be automatically generated for use with OziExplorer.

formats:
  - png                   # (default map output is in PNG and multi-layered TIFF format)
  - layered.tif

# Specify which layers to exclude from your map. This will prevent downloading of the layers and
# their inclusion in the final map. List each layer individually. Use the shortcuts 'coastal',
# 'UTM' and 'aerial' to exclude all coastal feature layers, UTM grid layers and aerial imagery
# layers, respectively.

exclude:
  - utm                   # (exclude UTM grid in favour of declination lines)
  - aerial-lpi-sydney     # (exclude hi-resolution sydney aerial imagery)
  - aerial-lpi-towns      # (exclude aerial imagery of towns and regional centres)

# Specify colours for individual topographic layers. Each colour should be specified as one of a
# recognised colour name (e.g. Red, Dark Magenta, Royal Blue), a quoted hex triplet (e.g. '#00FF00',
# '#2020e0', '#0033ff') or a decimal triplet (e.g. rgb(0,0,255), rgb(127,127,0)). The default color
# scheme closely matches the current 2nd-edition 25k NSW map sheets.
#
# N.B. If a photoshop (psd) file is being produced, do not use pure grayscale colours, as an
# ImageMagick bug will produce will produce a faulty layer in this case. Instead substitute a
# slight colour hue, e.g. #000001 instead of black, #808081 instead of middle-grey.

colours:
  watercourses: '#0033ff'         # blue for watercourses
  sand: '#ff6600'                 # red-brown for sand
  tracks-4wd: 'Dark Orange'       # dark orange for 4wd tracks
  tracks-vehicular: 'Dark Orange' # dark orange for vehicular tracs
  roads-unsealed: 'Dark Orange'   # dark orange for unsealed roads
  roads-sealed: 'Red'             # red for sealed roads
                                  # etc. etc.

# You can specify your own tiled pattern fills for various area layers if you really want to. Some
# of the default patterns are shown below. (This might be better done in Photoshop, however)

patterns:
  sand:                   # a diagonal dot pattern for sand
    01,10,01,00,00,00
    10,50,10,00,00,00
    01,10,01,00,00,00
    00,00,00,01,10,01
    00,00,00,10,50,10
    00,00,00,01,10,01
  orchards-plantations:   # a checker-board pattern for orchards, vineyards etc.
    111110000
    111110000
    111110000
    111110000
    111110000
    000000000
    000000000
    000000000
    000000000

Georeferencing
==============

The map projection used is transverse mercator, with a central meridian corresponding to the map's centre. This conformal projection is ideal for topographic maps. A grid for the nearest UTM zone (usually zone 55 or 56) can be added to the map (for use with a GPS) by including the UTM grid layers. All output layers (including the aerial imagery and shaded relief layers) are precisely aligned and in the same projection.

An associated world file (.wld) and proj4 projection file (.prj) are produced for the map. If you use Photoshop or Gimp to manually edit your map, you can use these files and the 'geotifcp' command to georeference your final map as a GeoTIFF:

  geotifcp -e map.wld -4 map.prj your-map.tif your-georeferenced-map.tif

If you choose png, gif or bmp as an output format, a .map file will also be produced for use with OziExplorer. (Note that if you move the map to a different location, you may need to edit the .map file to reflect the new location.)

Layer Description
=================

Topographic layers
These are the primary topographic features and cover all of NSW and the ACT. The data is the same as is used in the printed NSW topo series; however a key advantage is that they will likely include newer features (firetrails in particular) not present on the printed maps.

- vegetation: the base vegetation layer representing dense- and medium-crown forest; not particularly good quality and I recommend replacing this layer with one of your own, derived from an aerial imagery layer
- labels: contains labels for all roads, watercourse, contours, homesteads etc in black; these are all combined in one layer so as to avoid overlap of layers
- contours: regular and index contours (also contains dashed ancillary and hashed depression contours, if any exist), in dark magenta
- watercourses: watercourse lines, with single-pixel lines representing intermittent watercourses and thicker lines representing perennial watercourses, in blue
- water-areas: areas of water in rivers, lakes and larger dams, in light blue
- water-area-boundaries: boundaries of water areas, in blue
- water-areas-dry: normally-dry water areas, in dotted light blue
- water-areas-dry-boundaries: boundaries of normally-dry water areas, in blue
- dams: smaller farm dams, represented as blue squares
- ocean: ocean areas, in light blue
- coastline: ocean boundary, in black
- roads-sealed: sealed roads, represented as red lines with thicker lines representing distributor and arterial roads
- roads-unsealed: unsealed roads, represented as orange lines
- tracks-vehicular: unsealed vehicular tracks, represented as orange dashed lines
- tracks-4wd: 4wd tracks, represented as smaller  orange dashed lines
- pathways: various walking tracks, represented as thinner black dashed lines
- buildings: single buildings (e.g. homesteads), respresented as black squares
- intertidal: intertidal areas, in dotted blue
- inundation: land subject to inundation, in broken horizontal cyan lines
- reef: reef areas, in a cyan hash pattern
- rock-area: coastal rock areas, in a grey pattern
- sand: sand along rivers and beaches, in dotted brown pattern
- swamp-wet: wet swampy land, in cyan swamp pattern
- swamp-dry: dry swampy land, in brown swamp pattern
- cliffs: cliff sections, in grey bands
- clifftops: tops of said cliffs, as dotted pink line
- excavation: quarry faces, etc, as dotted gray line
- caves: caves and sinkholes, represented with small black icons
- rocks-pinnacles: large boulders, tors and rock pinnacles, respresented as pink stars
- built-up-areas: urban/residential areas, represented in light yellow
- pine: pine plantations, represented in dark green pine pattern
- orchards-plantations: orchards, vineyards and non-pine forest plantation, represented in green tile pattern
- building-areas: larger building complexes (e.g. shopping centres), represented in dark grey
- dam-walls: constructed dam walls, represented in black
- misc-perimeters: miscellaneous perimeters dividing different land use, represented as thin dashed gray lines
- towers: telecommuncation towers, etc, represented as small black icons
- mines: quarries and other mining areas, represented as small black icons
- yards: small stock yards, represented as black square outlines
- windmills: occasional farm windmills, represented as small black icons
- beacons: lighthouses and beacons, represented as small black icons
- railways: heavy- and light-gauge railway lines, represented as black hashed lines
- transmission-lines: high voltage electrical transmission lines, represented as black dot-dash lines
- landing-grounds: landing strips as found on farms, etc, represented as dark gray lines
- gates-grids: gates and grids on roads, represented as small black circular icons with two or one crossing line, respectively
- wharves: wharves and jetties, represented as black lines
- cadastre: NSW cadastral lines (property boundaries), represented as thin, light grey lines
- act-cadastre: ACT cadastral lines, represented as thin, light grey lines
- trig-points: trigonometric survey stations, represented as small black icons; not all trig points are present

Other topographic layers
These are various layers which are not included in the composite map, but may be useful in other ways or to fill in missing information in the NSW data.

- crown-reserves: areas of crown land
- act-rivers-and-creeks: watercourses, as derived from ACT map servers; not as good as the NSW equivalent
- act-urban-land: urban land, as derived from ACT map servers
- act-lakes-and-major-rivers: water areas, as derived from ACT map servers; not as good as the NSW equivalent
- act-plantations: pine plantations in ACT
- act-roads-sealed: sealed roads, as derived from ACT map servers
- act-roads-unsealed: unsealed roads, as derived from ACT map servers
- act-vehicular-tracks: vehicular tracks, as derived from ACT map servers
- act-adhoc-fire-access: various ad-hoc fire access, as derived from ACT map servers; may include walking tracks that are not represented in the NSW database

Aerial imagery
These are orthographic aerial images for the specified map area, derived from Google Maps, Nokia Maps, and the NSW LPI department. Depending on your map location there may be up to four different aerial images available.

These layers are very useful for confirming the accuracy of the topographic features. For example, you may be able to manually add firetrails, new dams, etc, which are missing from the NSW map layers, on the basis of what you can see in the aerial imagery. Since the images are correctly georeferenced, this is achieved simply by tracing out the extra information on the appropriate layer while viewing the aerial imagery underneath.

The other excellent use for these aerial imagery layers is to produce your own vegetation layer for a rogaine map. This can be accomplished using the 'color range' selection tool in Photoshop, for example, or other similar selection tools. (If you're feeling adventurous you can even try extracting a vegetation texture from the aerial image to emboss into your vegetation layer, imparting some lift to the map.)

Keep in mind that these aerial images have been warped into an orthographic projection from their original perspective, and may not always be pixel-perfect in alignment across the map area. They are still pretty good however, since we are typically in the 2 metre-per-pixel realm.

- aerial-lpi-ads40: the best, most recent high resolution imagery available from the NSW LPI; available for most but not all areas of interest
- aerial-lpi-sydney: high resolution imagery for the sydney area
- aerial-lpi-towns: medium-high resolution imagery for regional centres
- aerial-lpi-eastcoast: medium resolution imagery for most of the 25k topographic coverage; quite old imagery (from 90s?) and may be outdated
- aerial-google: generally good quality, recent aerial imagery from Google Maps; limited to 250 tiles per six hour period
- aerial-nokia: reasonable quality aerial imagery from Nokia Maps; limited to 250 tiles per six hours; georeferencing is not always the best and usually requires some manual nudging for best alignment

Annotation layers

- utm-grid: represents a UTM grid
- utm-eastings: annotates UTM eastings across the middle of the map
- utm-northings: annotates UTM northings across the middle of the map
- declination: represents lines of magnetic declination for map area
- control-numbers: represents control circles for rogaine courses
- control-circles: represents control numbers for rogaine courses

Elevation layers
These are grayscale images giving elevation and shaded-relief depictions for the map terrain. They are derived from the global ASTER digital elevation model (DEM), at about 45 metres per pixel.
TODO!!
- shaded-relief-315
- shaded-relief-45
- elevation



history.md
# Release History

* 12/12/2011: version 0.1 (initial release)
  * 13/12/2011: version 0.1.1: added bridges, floodways, fixed narrow gaps in roads
  * 14/12/2011: version 0.1.2: reworked UTM grid to display correctly across zone boundaries
* 21/12/2011: version 0.2: added map rotation; added specification of map bounds via gpx/kml file; added ability to auto-rotate map to minimise area.
* 11/01/2012: version 0.3: misc. additions (e.g. lookouts, campgrounds, rock/pinnacle labels, etc); collected point markers into single layer; separated permanent and intermittent water layers; prevented label/feature overlap; decreased download times; removed unavailable ACT layers; added low-res reference topo.
* 2/2/2012: version 0.4: added ferry routes, mangroves, restricted areas, canals, breakwaters, levees, road outlines; tweaked road & track colours; added grid-style UTM labels; removed absolute path from OziExplorer .map file; fixed bug wherein resolution tags in some output images were incorrectly set.
  * 8/2/2012: version 0.4.1: fixed bug whereby excluding labels also excluded control-labels
  * 9/2/2012: version 0.4.2: added kmz as output format
  * 13/2/2012: version 0.4.3: reworked road/track colours and outlines
  * 7/3/2012: version 0.4.4: fixed bug in OziExplorer .map files created by non-windows OS; added layer opacity; added overlay layers from GPX/KML/etc files
* 3/6/2012: version 0.5: Substantial rewrite to use the new NSW ArcGIS server
  * 4/6/2012: version 0.5.1: added metadata to identify layers correctly to Inkscape
* 30/6/2012: version 0.6: fixed bug with shaded-relief and vegetation layers on some versions of ImageMagick; added option for default nswtopo.cfg file stored in nswtopo.rb location; improved SVG topographic labels layer; added other output formats: .png, .tif, .kmz, .pdf, .prj, .wld, .map; added rendering using Inkscape or Batik; switched from 0.9996 to 1.0 for transverse mercator scale factor; changed config.yml to nswtopo.cfg; added configurations for individual output raster dpis and input raster resolutions
  * 5/7/2012: version 0.6.1: fixed vegetation rendering bug on linux; added time remaining estimations; bugfixes; added fix for java OutOfMemory error when using Batik
  * 5/8/2012: version 0.6.2: fixes to restore Windows compatibility and update Windows installation instructions
  * 4/10/2012: version 0.6.3: changed old LPIMAP layer names to new LPIMAP layer names; added the option of specifying a map bound using a track; fixed problem with ESRI SDS 1.95 1 font; fixed bug with KMZ generation; fixed broken cadastre layer; fixed broken holdings layer
  * 25/9/2013: version 0.6.4: fixed aerial-best, paths and holdings layers; expanded and renamed reference topo layers; updated vegetation layer to use v2 dataset.
* 10/2/2014: version 0.7: added in-place updating of composite map svg; added manual DEM option for shaded relief layer; store intermediate vegetation layer; added qlmanage and PhantomJS options for rasterising; added online source for 90m SRTM elevation data; added ability to import georeference raster images; added SPOT5 vegetation source; for rotated maps, prevent download of tiles which don't fall within map extents; scaled labels better for small-scale maps; added option to use 20-metre contour intervals; added option to exclude layers from map.
  * 22/2/2014: version 0.7.1: used all tracks instead of just first when calculating bounds from a GPX/KML file; fixed bug preventing tiny maps from downloading; changed manner of specifying rendering options; added alternate source of basic contour/road/track/watercourse/label layers; reverted to flat layer structure for SVG file; changed HydroArea layer to perennial water areas only; changed to LPIMapLocal as default data source due to availability.
* 3/7/2014: version 0.8: added RFS layers for stock dams and buildings; extracted various layer sources to external configuration files for greater flexibility; change way of specifying overlays; add ANC and water-drop icons for controls; add some SA and TAS map data sources.
  * 28/8/2014: version 0.8.1: change nsw/vegetation-2008-v2 woody vegetation colour; fix vegetation & relief rendering bug in Windows
  * 11/11/2014: version 0.8.2: add psd format output; fix bug in ArcGIS image servers; change to online source of 1-second SRTM relief data; apply median and bilateral filtering to shaded relief; fix label issues causing bad PhantomJS rasters; float vector label layers above feature layers.
* 27/1/2015: version 1.0: major update with NSW topographic maps now available for all areas and TAS topographic and vegetation layers also available.
  * 28/1/2015: version 1.0.1: fix 'command line is too long' bug on Windows.
* 1/8/2015: version 1.1: add features required for QLD maps; add version-checking; add TAS map features; fix some PhantomJS rendering bugs; misc tweaks/improvements/fixes.
  * 1/11/2015: version 1.1.1: add ability to render symbols for waypoint overlays; add fallback servers for unreliable map servers.
  * 13/11/2015: version 1.1.2: bugfixes for Windows and Ruby 1.9
  * 29/11/2015: version 1.1.3: bugfix for lambert conformal conic services
  * 13/12/2015: version 1.1.4: bugfix for Windows PhantomJS output
  * 5/2/2016: version 1.1.5: add template for NSW map sheets; update ACT server details; read gx:Track elements in KML files
  * 4/3/2016: version 1.1.6: bugfix for some changed NSW map servers
  * 18/6/2016: version 1.1.7: update NSW data sources
* 22/8/2016: version 1.2: improve labeling quality and speed; update various map sources accordingly
  * 25/8/2016: version 1.2.1: bugfix for stack overflow
* 23/9/2016: version 1.3: improvements to labelling algorithm; remove old NSW server references
* 5/10/2016: version 1.4: further labelling improvements; break out code into multiple files
  * 2/11/2016: version 1.4.1: add Electron as rasterising option; add multi-point shaded relief option; miscellaneous small fixes and refactoring

nswtopo.js
window.addEventListener('DOMContentLoaded', event => {
	document.querySelectorAll('details#avenza').forEach(details => {
		details.addEventListener('toggle', event => {
			const mapboxLoaded = new Promise(resolve => {
				const script = document.createElement('script');
				document.head.append(script);
				script.addEventListener('load', resolve, false);
				script.src = 'https://api.mapbox.com/mapbox.js/v3.3.1/mapbox.js';
			});
			const xhr = new XMLHttpRequest();
			const jsonLoaded = new Promise(resolve => {
				xhr.responseType = 'json';
				xhr.addEventListener('load', resolve, false);
				xhr.open('GET', 'maps.json'), xhr.send();
			});
			Promise.all([mapboxLoaded, jsonLoaded]).then(events => {
				if (xhr.status != 200) return;
				const sheets = xhr.response.features.map(feature => {
					return {
						type: feature['properties']['type'],
						url: feature['properties']['url'],
						state: feature['properties']['state'],
						title: feature['properties']['title'],
						corners: feature['geometry']['coordinates'][0].map(pair => pair.reverse()),
					};
				});
				const bounds = L.latLngBounds(sheets[0].corners);
				sheets.forEach(sheet => sheet.corners.forEach(point => bounds.extend(point)));
				L.mapbox.accessToken = 'pk.eyJ1IjoibWhvbGxpbmciLCJhIjoiY2pncms3d3plMDY3ODJ2bnh0YWdydTBwYyJ9.RdmqeL6b_5m8Q-SzQdbXuQ';
				const layer = L.mapbox.styleLayer('mapbox://styles/mholling/ck9bz5uth02671imzx2s7jujc');
				const map = L.mapbox.map('map').fitBounds(bounds).setMaxBounds(bounds.pad(0.2)).addLayer(layer);
				const div = details.querySelector('.map');
				const types = ['bundle', '50k', '40k', '25k'];
				const states = sheets.reduce((memo, sheet) => memo.add(sheet.state), new Set());
				types.forEach(type => states.forEach(state => map.createPane(type + ',' + state)));
				const addToggle = (toggle, klass) => {
					const label = document.createElement('label');
					const input = document.createElement('input');
					input.id = 'show-' + toggle;
					input.type = 'checkbox';
					input.checked = true;
					label.appendChild(input);
					label.append(toggle);
					div.querySelector('fieldset.' + klass).appendChild(label);
					label.querySelector('input').addEventListener('change', event => {
						Object.keys(map.getPanes()).forEach(key => {
							keys = key.split(',');
							if (!keys.includes(toggle))
								return;
							if (keys.every(key => div.querySelector('#show-' + key).checked))
								map.getPane(key).removeAttribute('hidden');
							else
								map.getPane(key).setAttribute('hidden', '');
						});
					});
				};
				types.forEach(type => addToggle(type, 'types'));
				Array.from(states).filter(state => !state.includes(',')).forEach(state => addToggle(state, 'states'));
				const qrCodeContainer = document.getElementById('qrcode');
				const qrCodeControl = L.control({position: 'bottomleft'});
				qrCodeControl.onAdd = map => qrCodeContainer;
				qrCodeControl.addTo(map);
				sheets.forEach(sheet => {
					const weight = sheet.type === 'bundle' ? 2 : 1;
					L.polygon(sheet.corners, {
						color: sheet.type === '25k' ? '#FF0000' : sheet.type === '40k' ? '#800080' : sheet.type === '50k' ? '#0000FF' : '#000000',
						weight: weight,
						opacity: 0.8,
						fillOpacity: 0.05,
						pane: sheet.type + ',' + sheet.state,
					}).on('click', event => {
						window.open(sheet.url);
					}).on('mouseover', event => {
						event.target.setStyle({weight: 4});
						if ('ontouchstart' in document.documentElement) return;
						new QRCode(qrCodeContainer, {text: sheet.url, width: 128, height: 128});
					}).on('mouseout', event => {
						event.target.setStyle({weight: weight});
						if ('ontouchstart' in document.documentElement) return;
						qrCodeContainer.innerHTML = null;
					}).bindTooltip(sheet.title, {
						direction: 'top',
						opacity: 0.9,
						sticky: true,
					}).addTo(map);
				});
			});
		}, {once: true});
	});

	const defaultTitle = document.title;
	document.querySelectorAll('details').forEach(details => {
		const hash = '#' + details.id;
		const title = "nswtopo | " + details.querySelector('summary').innerText;
		const others = document.querySelectorAll('details:not(' + hash + ')');
		if (hash == location.hash)
			details.setAttribute('open', '');
		window.addEventListener('hashchange', event => {
			if (hash == location.hash)
				details.setAttribute('open', '');
			else
				details.removeAttribute('open');
		});
		details.addEventListener('toggle', event => {
			const allClosed = !details.open && !document.querySelector('details[open]');
			if (details.open)
				others.forEach(other => other.removeAttribute('open'));
			if (details.open ? location.hash != hash : location.hash && allClosed)
				history.pushState(null, '', details.open ? location.pathname + hash : location.pathname);
			if (details.open)
				document.title = title;
			else if (allClosed)
				document.title = defaultTitle;
		});
	});

	document.querySelectorAll('a.obf').forEach(anchor => {
		anchor.appendChild(document.createTextNode(atob('aW5mb0Buc3d0b3BvLmNvbQ==')));
		anchor.setAttribute('href', atob('bWFpbHRvOmluZm9AbnN3dG9wby5jb20='));
	});

	document.querySelectorAll('div.carousel').forEach(div => {
		const changeSlide = (next, prev) => {
			if (next)
				var target = div.querySelector('input:checked + input');
			else if (prev)
				var target = div.querySelector('input:checked').previousElementSibling;
			if (target)
				target.checked = true;
		};
		let start = null;
		const slides = div.querySelector('ul.slides');
		slides.addEventListener('keydown', event => changeSlide("ArrowRight" == event.key, "ArrowLeft" == event.key));
		slides.addEventListener('touchstart', event => start = start ? null : event, {passive: true});
		slides.addEventListener('touchend', event => {
			for (; start; start = null) {
				const dx = event.changedTouches[0].screenX - start.changedTouches[0].screenX;
				const dy = event.changedTouches[0].screenY - start.changedTouches[0].screenY;
				if (Math.abs(dy) < Math.abs(dx))
					changeSlide(dx < 0, dx > 0);
			}
		}, {passive: true});
	});
});
